### ISSUE TITLE
X-Forwarded-Proto is ignored

### ISSUE THREAD

**Issue by TGithubbr** [nextcloud/server#22496]
### How to use GitHub

* Please use the 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to show that you are affected by the same issue.
* Please don't comment if you have no relevant information to add. It's just extra noise for everyone subscribed to this issue.
* Subscribe to receive notifications on status change and new comments. 

### Steps to reproduce
1. Set up an nginx reverse proxy which terminates ssl and forwards via http to nextcloud
2. write into configuration: proxy_set_header X-Forwarded-Proto $scheme
3. try to access nextcloud via the proxy

### Expected behaviour
Nextcloud should look at the X-Forwarded-Proto header and reply with the https website

### Actual behaviour
Nextcloud replies with the http website which leads to security policy errors inside the browser, rendering the site unusable. The browser's error: [unknown Refused to send form data to 'http://cloud.domainname/' because it violates the following Content Security Policy directive: "form-action 'self'".

NOTE: overwriteprotocol is NOT a solution to this problem, because it makes the site inaccessable via the local ip which prevents much faster local uploading. Instead the X-Forwarded-Proto header needs to be handled correctly.

### Server configuration

**Operating system: FreeNAS**

**Web server: Apache**

**Database: MySQL**

**PHP version: 7.4.8**

**Nextcloud version: 19.0.2** (see Nextcloud admin page)

**Updated from an older Nextcloud/ownCloud or fresh install: no**

**Where did you install Nextcloud from: FreeNAS plugins**

**Are you using external storage, if yes which one:** local/smb/sftp/... irrelevant

**Are you using encryption:** no

**Are you using an external user-backend, if yes which one:** no

**Comment by Dragon988** [comment]
> NOTE: overwriteprotocol is NOT a solution to this problem, because it makes the site **inaccessable from local network**

No it doesn't. Edit your hosts file to point the domain at your LAN ip.

**Comment by TGithubbr** [comment]
> > NOTE: overwriteprotocol is NOT a solution to this problem, because it makes the site **inaccessable from local network**
> 
> No it doesn't. Edit your hosts file to point the domain at your LAN ip.

sorry I meant via the local ip. Edited this

**Comment by Dragon988** [comment]
> sorry I meant via the local ip. Edited this.

True, but since the hosts file can be edited on windows and linux your only issue is on mobile devices... and there are workarounds for that as well.

Not to dismiss your issue, certainly worth looking into.

Cheers!

**Comment by Dragon988** [comment]
> This is lazy. Should I edit the host files of every computer including those where I don't have admin access?

No. If this bothers you then use then use the trusted_domains option which can be an array and have more then one entry.

Which enables both my LAN IP and domain.

**Comment by TGithubbr** [comment]
> > This is lazy. Should I edit the host files of every computer including those where I don't have admin access?
> 
> No. If this bothers you then use then use the trusted_domains option which can be an array and have more then one entry.
> 
> Which enables both my LAN IP and domain.

This is the solution! Thank you!

**Comment by huww98** [comment]
I'm wondering if there are any technical or security reasons that the `X-Forwarded-Proto` header cannot be supported? Or we can contribute on this? I think `X-Forwarded-Proto` is very commonly used and supporting this can simplify the config and reduce confusion.

**Comment by TGithubbr** [comment]
it is supported. you have to use the trusted domain option. For some reason this is not documented anywhere but adding the ip of the proxy enables x-forwarded-proto

**Comment by huww98** [comment]
I have set `trusted_proxies` and `trusted_domains`, but I still need to set `overwriteprotocol` to `https` to get login works. If I don't set `overwriteprotocol`, login will not redirect automatically, but if I refresh the page, I can log in successfully. And 6 images failed to load due to `http` in the URL. But most links work properly. Once I set `overwriteprotocol`, all the problems disappear. Maybe I still get some config wrong? I've used `phpinfo()` to verify that `X-Forwarded-Proto` is actually received.
